README.md
![Monster Maker Social Banner](./monster-maker-banner.jpg)

# Giffyglyph's Monster Maker

[![License: CC BY-NC-ND 4.0](https://img.shields.io/badge/License-CC%20BY--NC--ND%204.0-lightgrey.svg)](https://creativecommons.org/licenses/by-nc-nd/4.0/)
[![Support me on Patreon](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fshieldsio-patreon.vercel.app%2Fapi%3Fusername%3Dgiffyglyph%26type%3Dpatrons&style=flat-square)](https://patreon.com/giffyglyph)
[![Twitter](https://img.shields.io/twitter/follow/giffyglyph?color=%231DA1F2&style=flat-square)](http://twitter.com/giffyglyph)
[![Twitch](https://img.shields.io/twitch/status/giffyglyph?color=%23a45ee5&style=flat-square)](http://twitch.tv/giffyglyph)

Do you want to build fun, balanced monsters in D&D 5e? Is encounter design a huge timesink for you? Need some varied combat options that will get you running encounters in seconds? Then try **Giffyglyph's Monster Maker**: rules, advice, and templates to help you build new monsters and prepare combat encounters with ease!

* [Read the Latest PDF (v3.0.0)](https://github.com/giffyglyph/giffyglyphs-monster-maker/releases/download/v3.0.0/giffyglyphs_monster_maker_v3-0-0_en_pwyw.pdf)
* [Read the Changelog](https://github.com/giffyglyph/giffyglyphs-monster-maker/blob/master/CHANGELOG.md) 

## Features

* A fully-modular supplement that integrates easily with RAW 5e.
* Scrap Challenge Ratings for easy-to-use **Combat Levels**.
* Give your monsters some **Combat Ranks**: build Grunts, Minions, Elites, and Paragons.
* Specialize your fighting style with 6 **Combat Roles**: Controllers, Defenders, Lurkers, Skirmishers, Strikers, and Supporters.
* Build a specialist with 6 **Archetypes**: Spellcasters, Summoners, Swarms, Sidekicks, Hazards, and Lifelinkers.
* Act multiple times a round with **Paragon Power**.
* Equip your monster with 200+ categorized **Monstrous Traits**.
* Build new attacks easily with a step-by-step, fully-balanced **Feature & Effect Builder**.
* Build new encounters in seconds with **Threat**.
* Apply new status effects with **18 new conditions**.
* And tons more!

| Cover                                                    | Chapters (v3.0.0)                                           |
| -------------------------------------------------------- | ----------------------------------------------------------- |
| <img src="./monster-maker-cover.jpg" width="100%"> | <img src="./monster-maker-contents.jpg" width="100%"> |

## Roadmap

To see what's being worked on right now—and what may be planned for the future—check out the **[projects board](https://github.com/giffyglyph/giffyglyphs-monster-maker/projects)**.

## Bugs and Feature Suggestions

If you notice a bug or have a feature suggestion, visit the **[issue board](https://github.com/giffyglyph/giffyglyphs-monster-maker/issues)** and open a ticket. Please make sure to be as thorough as possible in your report and attach screenshots where appropriate, as low-effort tickets may be closed out-of-hand.

## Support

[![Support me on Patreon](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fshieldsio-patreon.vercel.app%2Fapi%3Fusername%3Dgiffyglyph%26type%3Dpatrons&style=flat-square)](https://patreon.com/giffyglyph)
[![Twitter](https://img.shields.io/twitter/follow/giffyglyph?color=%231DA1F2&style=flat-square)](http://twitter.com/giffyglyph)
[![Twitch](https://img.shields.io/twitch/status/giffyglyph?color=%23a45ee5&style=flat-square)](http://twitch.tv/giffyglyph)

If you'd like to see more from this project in future, please consider [becoming a patron](https://www.patreon.com/giffyglyph). You can also find more of my work at:

* [giffyglyph.com](https://giffyglyph.com)
* [twitter](https://twitter.com/giffyglyph)
* [twitch.tv/giffyglyph](https://twitch.tv/giffyglyph)

## Licensing

[![License: CC BY-NC-ND 4.0](https://img.shields.io/badge/License-CC%20BY--NC--ND%204.0-lightgrey.svg)](https://creativecommons.org/licenses/by-nc-nd/4.0/)

This work is licensed under a [Creative Commons Attribution-NonCommercial-NoDerivatives 4.0 International License](http://creativecommons.org/licenses/by-nc-nd/4.0/).


CHANGELOG.md
![Monster Maker Social Banner](./img/monster-maker-banner.jpg)

# Changelog

[![Support me on Patreon](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Fshieldsio-patreon.vercel.app%2Fapi%3Fusername%3Dgiffyglyph%26type%3Dpatrons&style=flat-square)](https://patreon.com/giffyglyph)
[![Twitter](https://img.shields.io/twitter/follow/giffyglyph?color=%231DA1F2&style=flat-square)](http://twitter.com/giffyglyph)
[![Twitch](https://img.shields.io/twitch/status/giffyglyph?color=%23a45ee5&style=flat-square)](http://twitch.tv/giffyglyph)

## v3.0.0
* Complete reboot and rewrite.
* Added 22 new chapters (too many changes to individually comment on).
* Combat ranks renamed: now Grunts, Minions, Elites, and Paragons.
* Added new elusive feature for minions.
* Removed the old Sniper combat role—we don't need a dedicated role for range when lurkers/strikers can fill that purpose.
* Added 3 new subroles for each Combat Role.
* Added 1 unique feature to each Combat Role.
* Added 6 new archetypes.
* Scalers now use threat as an advisory metric.
* ATK/DC bonuses are now relative (added to ability modifier).
* Math is readjusted; monster HP generally reduced by 20-50%.
* Brand new style for monster and encounter combat profiles.

## v2.1.2
* Updated PDF and website styles to match latest **Darker Dungeons** appearance.

## v2.1.0
* Complete style and layout revamp—prettier stat blocks, yay.
* Added 15+ new traits (p8-9)
* Added 20+ new powers (p10-11)
* Solo monster HP is now scales to number of players (reduce from players+1).
* Solo monster XP now scales with number of players (changed by x4).
* Assorted typos.
* New website grimoire format.

## v2.0.0
* Added spacer title page for correct page numbering when printing.
* Added 30+ new traits (p8-9)
* Clarified text on solo phases/transformations (p12).
* Moved Paragon Actions to a dedicated chapter (p13).
* Moved CR-MR converting to a dedicated chapter (p14-15).
* Added movement & resource attack types (p20).
* Added 11 example freeform attacks (p21).
* Added rules for summoning magic (p23).
* Added 10 example freeform spells (p24).
* New chapter: Attack Rarity (p25-26).
* New chapter: Overkill Attacks (p27-30).
* New adventure, the Court of Bone (p36-46)
* Clarified how to read/use monster templates (p48).

## v1.3.0
* Book-wide art/layout revision for improved readability.
* Added a chapter on randomised damage (p18-19).
* Revised monster profiles to display saves/abilities more clearly.

## v1.2.2
* Added notes on counterspell (p17).
* Added deadly overkill attacks (p17).
* Added a quick reference page (p58).

## v1.2.1
* Updated page art and layout.
* Added full contents page (p2).
* Added clarification to saving throws (p3).
* Added CR ratings the monster stat templates (p21-54).
* Added PDF bookmarks.

## v1.2.0
* Increased damage for elites (x1.1) and solos (x1.2).
* Increased hp for solos.
* Players can now spend one hit die during solo phase transitions.
* Reduce spell DCs and saving throws.
* Primary/secondary spell DCs now listed.
* Spell DCs and saving throws now affected by monster role.
* Added monster XP values for all monster types.
* Expanded templates with solo monsters for 3/4/5 players.
* Added CR-to-ML conversion notes.
* Merged elite/solo actions into Paragon Actions.

## v1.1.0

* Added templates for every monster level/type/role combination for extra-quick reference.
* Expanded freeform attacks to cover conditions, ongoing damage, and area attacks.
* Clarified healing magic and multiattacks.
* Added 2 new powers for each monster role.
* Added 20 new monster traits.
* Minions now have 20% of a standard monster's hit points, up from just 1hp.

## v1.0.0
* Added chapters: "Making Monsters", "Making Attacks", "Minions, Elites, & Solos", "Building an Encounter", "FAQ".


modules/projectManager.js
const diceDeviations = {
	4: [ 1.12, 1.58, 1.94, 2.24, 2.5, 2.74, 2.96, 3.16, 3.35, 3.54, 3.71, 3.87, 4.03, 4.18, 4.33, 4.47],
	6: [ 1.71, 2.42, 2.96, 3.42, 3.82, 4.18, 4.52, 4.83, 5.12, 5.40, 5.66, 5.92, 6.16, 6.39, 6.61, 6.83],
	8: [ 2.29, 3.24, 3.97, 4.58, 5.12, 5.61, 6.06, 6.48, 6.87, 7.25, 7.60, 7.94, 8.26, 8.57, 8.87, 9.17],
	10: [ 2.87, 4.06, 4.97, 5.74, 6.42, 7.04, 7.60, 8.12, 8.62, 9.08, 9.53, 9.95, 10.36, 10.75, 11.12, 11.49],
	12: [ 3.45, 4.88, 5.98, 6.90, 7.72, 8.46, 9.13, 9.76, 10.36, 10.92, 11.45, 11.96, 12.45, 12.92, 13.37, 13.81],
	20: [ 5.77, 8.15, 9.99, 11.53, 12.89, 14.12, 15.26, 16.31, 17.30, 18.23, 19.12, 19.97, 20.79, 21.58, 22.33, 23.07],
	100: [ 28.87, 40.82, 50.00, 57.73, 64.55, 70.71, 76.37, 81.65, 86.60, 91.28, 95.74, 99.99, 104.08, 108.01, 111.80, 115.46]
};

/**
 * Apply page numbers and chapters.
 * @param {Object} dom - The dom.
 * @param {boolean} isCollection - True if the dom represents a collection.
 * @returns {Object} The dom.
 */
function processDom(job, dom) {
	dom.querySelectorAll('.page__header .page__book .title').forEach((x) => x.innerHTML = "Giffyglyph's Monster Maker");
	dom.querySelectorAll('.table--monster-statistics').forEach(function (element) {
		let rows = "";
		for (let i = (element.dataset.levelStart ? element.dataset.levelStart : 0); i <= (element.dataset.levelEnd ? element.dataset.levelEnd : 20); i++) {
			const rank = element.dataset.rank ? element.dataset.rank : "grunt";
			const threat = element.dataset.threat ? element.dataset.threat : 1;
			const template = job.format.module.scalerFactory.createScalerTemplate(i, rank, null, threat);
      rows += `
        <tr>
          <td>${template.combatLevel}</td>
          <td>${template.abilityModifiers.map((x) => _renderNumber(x.getValue())).join(", ")}</td>
          <td>${_renderNumber(template.proficiencyBonus.getValue())}</td>
          <td>${template.initiativeBonus.getValue() ? _renderNumber(template.initiativeBonus.getValue()) : "—"}</td>
          <td>${template.armorClass.getValue()}</td>
          <td>${template.hitPoints.getValue().toLocaleString()}</td>
          <td>${template.trainedSavingThrows.getValue()}</td>
          <td>${_renderNumber(template.attackBonus.getValue())}</td>
          <td>${_renderNumber(template.attackDc.getValue())}</td>
          <td>${template.attackDamage.getValue()}</td>
          <td>${_renderChallengeRating(template.challengeRating.getValue())}</td>
          <td>${template.experiencePoints.getValue().toLocaleString()}</td>
        </tr>
      `;
    }
    element.innerHTML = `
      <colgroup>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
      </colgroup>
      <thead>
        <tr>
          <th rowspan=2>Level</th>
          <th colspan=3><span class="spacer">Attributes</span></th>
          <th colspan=3><span class="spacer">Defence</span></th>
          <th colspan=3><span class="spacer">Offense</span></th>
          <th colspan=2><span class="spacer">Ratings</span></th>
        </tr>
        <tr>
          <th>Ability Modifiers</th>
          <th>PB</th>
          <th>INIT</th>
          <th>AC</th>
          <th>HP</th>
          <th>TST</th>
          <th>ATK</th>
          <th>DC</th>
          <th>DMG</th>
          <th>CR</th>
          <th>XP</th>
        </tr>
      </thead>
      <tbody>${rows}</tbody>
    `;
	});
	dom.querySelectorAll('.table--scaler-templates').forEach(function (element) {
		const ranks = [ "minion:1", "grunt:1", "elite:1", "paragon:3", "paragon:4", "paragon:5", "paragon:6", "paragon:7" ];
		const roles = [ "controller", "defender", "lurker", "skirmisher", "striker", "supporter" ];
		let rows = "";
		ranks.forEach((x) => {
			let rank = x.split(':')[0];
			let players = x.split(':')[1];
			let stats = job.format.module.scalerFactory.createScalerTemplate(Number(element.dataset.level), rank, null, Number(players));
			rows += `
				<tr>
					<th colspan=21 class="header">
						<span>${job.translator.t(`scaler.rank.${rank}`, { "threat": players })}</span>
						<span class="info">Level ${element.dataset.level}</span>
						<span class="info">Ability Modifiers: ${stats.abilityModifiers.map((y) => _renderNumber(y.getValue())).join(", ")}</span>
            <span class="info">Threat: ${stats.combatThreat}</span>
					</th>
				</tr>
			`;
      roles.forEach((y) => {
				let level = job.format.module.scalerFactory.createScalerTemplate(Number(element.dataset.level), rank, y, Number(players));
				rows += `
					<tr>
						<td>${job.translator.t(`scaler.role.${y}`)}</td>
            <td>${level.proficiencyBonus.getValue() ? _renderNumber(level.proficiencyBonus.getValue()) : "—"}</td>
            <td>${level.initiativeBonus.getValue() ? _renderNumber(level.initiativeBonus.getValue()) : "—"}</td>
						<td>${(level.baseSpeed.getValue() - 30 != 0) ? _renderNumber(level.baseSpeed.getValue() - 30) : '—'}</td>
						<td>${level.armorClass.getValue()}</td>
						<td>${level.hitPoints.getValue().toLocaleString()}</td>
						<td>${level.trainedSavingThrows.getValue()}</td>
            <td>${_renderNumber(level.attackBonus.getValue())}</td>
            <td>${_renderNumber(level.attackDc.getValue())}</td>
						<td>${level.attackDamage.getValue()}</td>
            <td>${(level.paragonPower.getValue()) ? level.paragonPower.getValue() : "—"}</td>
            <td>${(level.paragonDefence.getValue()) ? level.paragonDefence.getValue() : "—"}</td>
						<td>${level.trainedSkill ? job.translator.t(`scaler.skill.${level.trainedSkill}`) : "—"}</td>
            <td>${_renderChallengeRating(stats.challengeRating.getValue())}</td>
						<td>${stats.experiencePoints.getValue().toLocaleString()}</td>
					</tr>
				`;
			});
		});
		
		element.innerHTML = `
      <colgroup>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
        <col>
      </colgroup>
			<thead>
				<tr>
					<th rowspan=2>Role</th>
          <th colspan=3><span class="spacer">Attributes</span></th>
          <th colspan=3><span class="spacer">Defence</span></th>
					<th colspan=3><span class="spacer">Offense</span></th>
          <th colspan=2><span class="spacer">Paragon</span></th>
					<th rowspan=2>Skill</th>
          <th colspan=2><span class="spacer">Ratings</span></th>
				</tr>
				<tr>
          <th>PB</th>
          <th>INIT</th>
          <th>SPD</th>
          <th>AC</th>
          <th>HP</th>
          <th>TST</th>
					<th>ATK</th>
          <th>DC</th>
					<th>DMG</th>
          <th>POW</th>
					<th>DEF</th>
          <th>CR</th>
					<th>XP</th>
				</tr>
			</thead>
			<tbody>${rows}</tbody>
		`;
	});

	dom.querySelectorAll('.table--random-damage [data-die]').forEach(function (element) {
		let mean = ((1 + Number(element.dataset.die)) / 2) * element.dataset.count;
		let deviation = diceDeviations[element.dataset.die][element.dataset.count - 1];
		element.outerHTML = `
			<tr>
				<td>${element.dataset.count}d${element.dataset.die}</td>
				<td>${Math.floor(mean)}</td>
				<td>${Math.floor(mean - deviation)}-${Math.ceil(mean + deviation)}</td>
				<td>${element.dataset.count * element.dataset.die}</td>
			</tr>
		`;
	});

  dom.querySelectorAll('.table--combat-stats [data-level]').forEach(function (element) {
    let stats = job.format.module.scalerFactory.createScalerTemplate(Number(element.dataset.level), "grunt", null, 1);
    element.outerHTML = `
      <tr>
        <td>${stats.combatLevel}</td>
        <td>${13 + (Math.floor(stats.combatLevel / 3))}</td>
        <td>${16 + (stats.combatLevel * 7)}</td>
        <td>${_renderNumber(Math.floor(stats.combatLevel / 3) - 1)}</td>
        <td>${_renderNumber(3 + Math.floor(stats.combatLevel / 2))}</td>
        <td>${11 + Math.floor(stats.combatLevel / 2)}</td>
        <td>${Math.max(1, (stats.combatLevel * 4) - 2)}</td>
        <td>${_renderNumber(stats.proficiencyBonus.getValue())}</td>
        <td>${_renderChallengeRating(stats.challengeRating.getValue())}</td>
        <td>${stats.experiencePoints.getValue().toLocaleString()}</td>
      </tr>
    `;
  });
		
	return dom;
}

function _renderChallengeRating(cr) {
	switch (cr) {
		case 0.125:
			return "1/8";
		case 0.25:
			return "1/4";
		case 0.5:
			return "1/2";
		default:
			return cr;
	}
}

function _renderNumber(number, hidePositive) {
	const positive = (hidePositive) ? '' : '+';
	return (number >= 0) ? positive + number : `&minus;${Math.abs(number)}`;
}

export { processDom };

src/modules/monster_maker.js
import * as frankenstein from './frankenstein.js';

const DICE = {
	d4: {
		description: 'Accurate',
		values: [
			{
				min: 1,
				max: 4,
				mean: 2.5,
				deviation: 1.12
			}, {
				min: 2,
				max: 8,
				mean: 5,
				deviation: 1.58
			}, {
				min: 3,
				max: 12,
				mean: 7.5,
				deviation: 1.94
			}, {
				min: 4,
				max: 16,
				mean: 10,
				deviation: 2.24
			}, {
				min: 5,
				max: 20,
				mean: 12.5,
				deviation: 2.5
			}, {
				min: 6,
				max: 24,
				mean: 15,
				deviation: 2.74
			}
		]
	},
	d6: {
		description: 'Predictable',
		values: [
			{
				min: 1,
				max: 6,
				mean: 3.5,
				deviation: 1.71
			}, {
				min: 2,
				max: 12,
				mean: 7,
				deviation: 2.42
			}, {
				min: 3,
				max: 18,
				mean: 10.5,
				deviation: 2.96
			}, {
				min: 4,
				max: 24,
				mean: 14,
				deviation: 3.42
			}, {
				min: 5,
				max: 30,
				mean: 17.5,
				deviation: 3.82
			}, {
				min: 6,
				max: 36,
				mean: 21,
				deviation: 4.18
			}
		]
	},
	d8: {
		description: 'Reliable',
		values: [
			{
				min: 1,
				max: 8,
				mean: 4.5,
				deviation: 2.29
			}, {
				min: 2,
				max: 16,
				mean: 9,
				deviation: 3.24
			}, {
				min: 3,
				max: 24,
				mean: 13.5,
				deviation: 3.97
			}, {
				min: 4,
				max: 32,
				mean: 18,
				deviation: 4.58
			}, {
				min: 5,
				max: 40,
				mean: 22.5,
				deviation: 5.12
			}, {
				min: 6,
				max: 48,
				mean: 27,
				deviation: 5.61
			}
		]
	},
	d10: {
		description: 'Irregular',
		values: [
			{
				min: 1,
				max: 10,
				mean: 5.5,
				deviation: 2.87
			}, {
				min: 2,
				max: 20,
				mean: 11,
				deviation: 4.06
			}, {
				min: 3,
				max: 30,
				mean: 16.5,
				deviation: 4.97
			}, {
				min: 4,
				max: 40,
				mean: 22,
				deviation: 5.74
			}, {
				min: 5,
				max: 50,
				mean: 27.5,
				deviation: 6.42
			}, {
				min: 6,
				max: 60,
				mean: 33,
				deviation: 7.04
			}
		]
	},
	d12: {
		description: 'Volatile',
		values: [
			{
				min: 1,
				max: 12,
				mean: 6.5,
				deviation: 3.45
			}, {
				min: 2,
				max: 24,
				mean: 13,
				deviation: 4.88
			}, {
				min: 3,
				max: 36,
				mean: 19.5,
				deviation: 5.98
			}, {
				min: 4,
				max: 48,
				mean: 26,
				deviation: 6.9
			}, {
				min: 5,
				max: 60,
				mean: 32.5,
				deviation: 7.72
			}, {
				min: 6,
				max: 72,
				mean: 39,
				deviation: 8.46
			}
		]
	},
	d20: {
		description: 'Wild',
		values: [
			{
				min: 1,
				max: 20,
				mean: 10.5,
				deviation: 5.77
			}, {
				min: 2,
				max: 40,
				mean: 21,
				deviation: 8.15
			}, {
				min: 3,
				max: 60,
				mean: 31.5,
				deviation: 9.99
			}, {
				min: 4,
				max: 80,
				mean: 42,
				deviation: 11.53
			}, {
				min: 5,
				max: 100,
				mean: 52.5,
				deviation: 12.89
			}, {
				min: 6,
				max: 120,
				mean: 63,
				deviation: 14.12
			}
		]
	},
	d100: {
		description: 'Chaotic',
		values: [
			{
				min: 1,
				max: 100,
				mean: 50.5,
				deviation: 28.87
			}, {
				min: 2,
				max: 200,
				mean: 101,
				deviation: 40.82
			}, {
				min: 3,
				max: 300,
				mean: 151.5,
				deviation: 50
			}, {
				min: 4,
				max: 400,
				mean: 202,
				deviation: 57.73
			}, {
				min: 5,
				max: 500,
				mean: 252.5,
				deviation: 64.55
			}, {
				min: 6,
				max: 600,
				mean: 303,
				deviation: 70.71
			}
		]
	}
};

/**
 * Apply project-specific functions
 * @param {object} html DOM
 * @param {string} format output format
 * @param {string} language output language
 * @return {object} html DOM
 */
export function apply (html, format, language) {
	frankenstein.createMonsters(html);
	frankenstein.createMonsterInfo(html);

	html.querySelectorAll('.table--monster-statistics tbody').forEach(function (element) {
		for (let i = -5; i <= 35; i++) {
			const level = frankenstein.getStatsForLevel(i);
			element.innerHTML += `
				<tr>
					<td>${frankenstein.formatNumber(level.level, true)}</td>
					<td>${level.ac}</td>
					<td>${level.hp}</td>
					<td>${frankenstein.formatNumber(level.attack)}</td>
					<td>${level.damage}</td>
					<td>${level.dc}, ${(level.dc - 3)}</td>
					<td>${frankenstein.formatNumber(level.skills)}</td>
					<td>${frankenstein.formatNumber(level.proficiency)}</td>
					<td>${frankenstein.formatSaves(level.saves, true)}</td>
					<td>${frankenstein.formatAbilities(level.abilities, true)}</td>
					<td>${Math.floor(level.xp / 4).toLocaleString()}</td>
				</tr>
			`;
		}
	});

	html.querySelectorAll('.table--monster-roles tbody').forEach(function (element) {
		const roles = frankenstein.getRoles().sort(function (a, b) {
			return a.name.localeCompare(b.name);
		});

		roles.forEach(function (role) {
			element.innerHTML += `
				<tr>
					<td>${role.name}</td>
					<td>${role.ac != 0 ? frankenstein.formatNumber(role.ac) : `<span class='empty'>—</span>`}</td>
					<td>${role.saves != 0 ? frankenstein.formatNumber(role.saves) : `<span class='empty'>—</span>`}</td>
					<td>${role.hp != 100 ? `x ${role.hp / 100}` : `<span class='empty'>—</span>`}</td>
					<td>${role.attack != 0 ? frankenstein.formatNumber(role.attack) : `<span class='empty'>—</span>`}</td>
					<td>${role.damage != 100 ? `x ${role.damage / 100}` : `<span class='empty'>—</span>`}</td>
					<td>${role.speed != 0 ? frankenstein.formatNumber(role.speed) : `<span class='empty'>—</span>`}</td>
					<td>${role.perception ? `Trained` : `<span class='empty'>—</span>`}</td>
					<td>${role.stealth ? `Trained` : `<span class='empty'>—</span>`}</td>
					<td>${role.initiative ? `Trained` : `<span class='empty'>—</span>`}</td>
				</tr>
			`;
		});
	});

	html.querySelectorAll('.table--monster-traits tbody').forEach(function (element) {
		const traits = frankenstein.getTraits().sort(function (a, b) {
			return a.name.localeCompare(b.name);
		});
		const parent = element.closest('table');
		const start = parent.getAttribute('data-start-from') ? Number(parent.getAttribute('data-start-from')) : 1;
		const end = parent.getAttribute('data-end-at') ? Number(parent.getAttribute('data-end-at')) : traits.length;

		for (let i = (start - 1); i < end; i++) {
			const trait = traits[i];
			if (trait) {
				element.innerHTML += `
					<tr>
						<td>${i + 1}</td>
						<td>${trait.name}</td>
						<td>${frankenstein.removeShortcodes(trait.description)}</td>
					</tr>
				`;
			}
		}
	});

	html.querySelectorAll('.table--monster-powers tbody').forEach(function (element) {
		const roles = frankenstein.getRoles().sort(function (a, b) {
			return a.name.localeCompare(b.name);
		});
		const parent = element.closest('table');
		const start = parent.getAttribute('data-start-from') ? Number(parent.getAttribute('data-start-from')) : 1;
		const end = parent.getAttribute('data-end-at') ? Number(parent.getAttribute('data-end-at')) : (roles.map((x) => x.powers).reduce((acc, list) => acc + list.length, 0) + 1);
		let i = 1;

		roles.forEach(function (role) {
			if (i >= start && i <= end) {
				element.innerHTML += `
					<tr class='header'>
						<td colspan='3'><span>${role.name}</span></td>
					</tr>
				`;
			}

			const powers = role.powers.sort(function (a, b) {
				return a.name.localeCompare(b.name);
			});
			powers.forEach(function (power, index) {
				if (i >= start && i <= end) {
					element.innerHTML += `
						<tr>
							<td>${i}</td>
							<td>${power.name}</td>
							<td>${frankenstein.removeShortcodes(power.description)}</td>
						</tr>
					`;
				}
				i++;
			});
		});
	});

	html.querySelectorAll('.table--random-damage tbody').forEach(function (element) {
		[ 'd4', 'd6', 'd8', 'd10', 'd12', 'd20', 'd100' ].forEach(function (type) {
			const data = DICE[type];

			element.innerHTML += `
				<tr class='header'>
					<td colspan='6'><span>${data.description}</span></td>
				</tr>
			`;

			data.values.forEach(function (value, index) {
				element.innerHTML += `
					<tr>
						<td>${index + 1}${type}</td>
						<td>${value.min}-${value.max}</td>
						<td>${value.mean}</td>
						<td>${value.deviation}</td>
						<td>${Math.floor(value.mean - value.deviation)}-${Math.ceil(value.mean + value.deviation)}</td>
						<td>${(index + 1)}${type} + [static damage − ${Math.floor(value.mean)}]</td>
					</tr>
				`;
			});
		});
	});

	html.querySelectorAll('list[data-type=\'monster-templates\']').forEach(function (element) {
		const level = Number(element.getAttribute('data-level'));
		const templateTypes = [ 'Minion:1', 'Standard:1', 'Elite:1', 'Solo:3', 'Solo:4', 'Solo:5', 'Solo:6' ];
		const templateRoles = [ 'Controller', 'Defender', 'Lurker', 'Scout', 'Sniper', 'Striker', 'Supporter' ];
		let content = ``;

		templateTypes.forEach(function (type) {
			const tokens = type.split(':');

			templateRoles.forEach(function (role, index) {
				const players = tokens[1];
				const monster = frankenstein.getMonsterStats({
					name: `${type} ${role}`,
					level: level,
					role: role,
					type: tokens[0],
					players: players
				});

				content += `
					<tr ${index == 0 ? `class='section'` : ``}>
						<td>${(tokens[1] > 1) ? `${tokens[0]} vs ${tokens[1]}` : tokens[0]}</td>
						<td>${role}</td>
						<td>${monster.stats.ac}</td>
						<td>${monster.stats.hp.max.toLocaleString()}</td>
						<td>${(tokens[0] == `Solo`) ? `3 x ${monster.stats.hp.phase}` : `2 x ${monster.stats.hp.bloodied}`}</td>
						<td>${frankenstein.formatNumber(monster.stats.attack)}</td>
						<td>${monster.stats.damage}</td>
						<td>${monster.stats.dc.primary}, ${monster.stats.dc.secondary}</td>
						<td>${frankenstein.formatNumber(monster.stats.perception, true)}</td>
						<td>${frankenstein.formatNumber(monster.stats.initiative, true)}</td>
						<td>${frankenstein.formatNumber(monster.stats.stealth, true)}</td>
						<td>${frankenstein.formatSaves(monster.stats.saves, true)}</td>
						<td>${monster.cr}</td>
						<td>${monster.xp}</td>
					</tr>
				`;
			});
		});

		element.innerHTML += `
			<header class="panel__header">
				<h4 class="panel__title">Monster Templates: Level ${level}</h4>
			</header>
			<section class="panel__body">
				<table class='table table--monster-templates'>
					<thead>
						<tr>
							<th>Rank</th>
							<th>Role</th>
							<th>AC</th>
							<th colspan=2><div class='spacer'>Hit Points</div></th>
							<th>Atk Bonus</th>
							<th>Dmg / Action</th>
							<th>Spell DCs</th>
							<th>Perc.</th>
							<th>Init.</th>
							<th>Stl.</th>
							<th>Saving Throws</th>
							<th>CR</th>
							<th>XP</th>
						</tr>
						<tr>
							<th colspan=3></th>
							<th>Max</th>
							<th>Phases</th>
							<th colspan=9></th>
						</tr>
					</thead>
					<tbody>
						${content}
					</tbody>
					<tfoot>
						<tr>
							<th colspan=3></th>
							<th>Max</th>
							<th>Phases</th>
							<th colspan=9></th>
						</tr>
						<tr>
							<th>Rank</th>
							<th>Role</th>
							<th>AC</th>
							<th colspan=2><div class='spacer'>Hit Points</div></th>
							<th>Atk Bonus</th>
							<th>Dmg / Action</th>
							<th>Spell DCs</th>
							<th>Perc.</th>
							<th>Init.</th>
							<th>Stl.</th>
							<th>Saving Throws</th>
							<th>CR</th>
							<th>XP</th>
						</tr>
					</tfoot>
				</table>
			</section>
		`;
	});

	return html;
}
